### PR TITLE
Add option to log but not enforce connection filtering, reset proposal stats through Admin server API 

### PR DESCRIPTION
**Pull request by abhilash1in**
### Description
1. Add option to log but not enforce connection filtering.
    - When ZK server is in dedicated mode and  `X509ZNodeGroupAclProvider.enforceDedicatedDomain` is **false** and a client with ID not part of the corresponding dedicated domain tries to connect, ZK server logs a warning but allows the connection to be established. When `X509ZNodeGroupAclProvider.enforceDedicatedDomain` is **true**, ZK server disconnects the connection (current and default behavior).
2. Reset proposal stats through Admin server API (in line with [stat reset 4lw command](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/server/command/StatResetCommand.java#L39))
3. [**Update: These changes were removed from this PR and will be incorporated in a different PR** ] Removes misconfigured attempts to parse `zookeeper.X509ZNodeGroupAclProvider` related properties from zoo.cfg.
    - Let's take an example `zoo.cfg` line, `X509ZNodeGroupAclProvider.setX509ClientIdAsAcl=true`
    - This will not be captured by
       ```
       if (key.equals(X509AuthenticationConfig.SET_X509_CLIENT_ID_AS_ACL)) {
                X509AuthenticationConfig.getInstance().setX509ClientIdAsAclEnabled(value);
       } 
       ```
    - This is because `X509AuthenticationConfig.SET_X509_CLIENT_ID_AS_ACL` is equal to `zookeeper.X509ZNodeGroupAclProvider.setX509ClientIdAsAcl`, but `key` is `X509ZNodeGroupAclProvider.setX509ClientIdAsAcl` (without `zookeeper.`).
    - The zoo.cfg parsing of `X509ZNodeGroupAclProvider` and usage of associated setters (like in the above `if` block) was never being used so far.
    - Instead, we were relying on the default `else` block below to set `X509ZNodeGroupAclProvider.setX509ClientIdAsAcl` from zoo.cfg as `zookeeper.X509ZNodeGroupAclProvider.setX509ClientIdAsAcl` System property, which is then read by associated getters.
      ```
       else {
           System.setProperty("zookeeper." + key, value);
       }
      ```
    
### Tests
```
mvn test -e -Dtest=X509ZNodeGroupAclProviderTest#testConnectionFiltering -DfailIfNoTests=false
```

Dedicated domain = `DomainX`

1. Authorized domain is allowed to connect
```
2023-05-12 16:40:16,314 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@76] - Client URI domain mapping root path: /zookeeper/uri-domain-map
2023-05-12 16:40:16,317 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: DomainYUser --> DomainY
2023-05-12 16:40:16,317 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: CrossDomainUser --> CrossDomain
2023-05-12 16:40:16,318 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: DomainXUser --> DomainX
2023-05-12 16:40:16,318 [myid:] - INFO  [main:X509ZNodeGroupAclProvider@194] - New UriDomainMappingHelper has been instantiated.
2023-05-12 16:40:16,322 [myid:] - WARN  [main:X509AuthenticationUtil@74] - Key store location not specified for SSL
2023-05-12 16:40:16,322 [myid:] - WARN  [main:X509AuthenticationUtil@106] - Truststore location not specified for SSL
2023-05-12 16:40:16,322 [myid:] - INFO  [main:X509AuthenticationConfig@245] - zookeeper.X509ZNodeGroupAclProvider.dedicatedDomain = DomainX
2023-05-12 16:40:16,322 [myid:] - INFO  [main:X509ZNodeGroupAclProvider@237] - Id 'DomainXUser' belongs to domain that matches server namespace 'DomainX', authorized for access.
2023-05-12 16:40:16,323 [myid:] - INFO  [main:X509ZNodeGroupAclProvider@277] - Authenticated Id 'scheme: x509, id: DomainXUser' has been added to session 0x0 from host .
```

2. Unauthorized domain is not allowed to connect (connection filtering is enforced)
```
2023-05-12 16:40:16,323 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@76] - Client URI domain mapping root path: /zookeeper/uri-domain-map
2023-05-12 16:40:16,324 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: DomainYUser --> DomainY
2023-05-12 16:40:16,324 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: CrossDomainUser --> CrossDomain
2023-05-12 16:40:16,324 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: DomainXUser --> DomainX
2023-05-12 16:40:16,324 [myid:] - INFO  [main:X509ZNodeGroupAclProvider@194] - New UriDomainMappingHelper has been instantiated.
2023-05-12 16:40:16,324 [myid:] - INFO  [main:X509AuthenticationConfig@245] - zookeeper.X509ZNodeGroupAclProvider.dedicatedDomain = DomainY
2023-05-12 16:40:16,324 [myid:] - INFO  [main:X509AuthenticationConfig@250] - zookeeper.X509ZNodeGroupAclProvider.enforceDedicatedDomain = true
2023-05-12 16:40:16,325 [myid:] - ERROR [main:X509ZNodeGroupAclProvider@243] - Id 'DomainXUser' does not belong to domain that matches server namespace 'DomainY', disconnecting the connection.
```

3. Unauthorized domain is allowed to connect with a warning log (connection filterning not enforced)
```
2023-05-12 16:40:16,325 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@76] - Client URI domain mapping root path: /zookeeper/uri-domain-map
2023-05-12 16:40:16,325 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: DomainYUser --> DomainY
2023-05-12 16:40:16,325 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: CrossDomainUser --> CrossDomain
2023-05-12 16:40:16,325 [myid:] - INFO  [main:ZkClientUriDomainMappingHelper@135] - Registering client URI domain mapping: DomainXUser --> DomainX
2023-05-12 16:40:16,325 [myid:] - INFO  [main:X509ZNodeGroupAclProvider@194] - New UriDomainMappingHelper has been instantiated.
2023-05-12 16:40:16,326 [myid:] - INFO  [main:X509AuthenticationConfig@245] - zookeeper.X509ZNodeGroupAclProvider.dedicatedDomain = DomainY
2023-05-12 16:40:16,326 [myid:] - INFO  [main:X509AuthenticationConfig@250] - zookeeper.X509ZNodeGroupAclProvider.enforceDedicatedDomain = false
2023-05-12 16:40:16,326 [myid:] - WARN  [main:X509ZNodeGroupAclProvider@247] - Id 'DomainXUser' does not belong to domain that matches server namespace 'DomainY'. Allowing connection for now, but client connection will be blocked/disconnected in future when dedicated domain connection filtering is enforced.
```

### CODE REVIEW

File: zookeeper-client/zookeeper-client-c/src/zookeeper.c [hunk 1]
```diff
@@ -2592,11 +2592,6 @@ static int init_ssl_for_socket(zsock_t *fd, zhandle_t *zh, int fail_on_error) {
         OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
         method = TLS_client_method();
 #endif
-        if (FIPS_mode() == 0) {
```
Comment (desaikomal) [inline zookeeper-client/zookeeper-client-c/src/zookeeper.c:2595]: nit : just a suggestion - code clean up can be done separately from the actual code change. j

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/admin/Commands.java [hunk 2]
```diff
@@ -613,7 +614,11 @@ public StatResetCommand() {
         @Override
         public CommandResponse run(ZooKeeperServer zkServer, Map<String, String> kwargs) {
             CommandResponse response = initializeResponse();
-            zkServer.serverStats().reset();
+            ServerStats serverStats = zkServer.serverStats();
```
Comment (desaikomal) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/admin/Commands.java]: nit:  is there any reason, why we needed to split this into 2 lines? you are not using 'serverStats' any other place AFAIK.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java [hunk 3]
```diff
@@ -330,6 +353,13 @@ public String getZnodeGroupAclServerDedicatedDomain() {
     return znodeGroupAclServerDedicatedDomain;
   }
 
+  public boolean shouldEnforceDedicatedDomain() {
+    if (znodeGroupAclServerShouldEnforceDedicatedDomain == null) {
+      setZnodeGroupAclServerShouldEnforceDedicatedDomain(System.getProperty(ENFORCE_DEDICATED_DOMAIN, "true"));
```
Comment (desaikomal) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java:382]: nit: you can directly use your instance variable?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java [hunk 4]
```diff
@@ -330,6 +353,13 @@ public String getZnodeGroupAclServerDedicatedDomain() {
...
+    }
+    return Boolean.parseBoolean(znodeGroupAclServerShouldEnforceDedicatedDomain);
```
Comment (desaikomal) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java:384]: should you do with ignoreCase ?? just in case, user types TRUE vs/vs true

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeerConfig.java [hunk 5]
```diff
@@ -377,28 +376,6 @@ public void parseProperties(Properties zkProp) throws IOException, ConfigExcepti
                 backupConfigBuilder.setTimetableStoragePath(value);
             } else if (key.equals(BackupSystemProperty.BACKUP_TIMETABLE_BACKUP_INTERVAL_MS)) {
                 backupConfigBuilder.setTimetableBackupIntervalInMs(Long.parseLong(value));
-            } else if (key.equals(X509AuthenticationConfig.SET_X509_CLIENT_ID_AS_ACL)) {
```
Comment (desaikomal) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeerConfig.java:380]: any reason why you are deleting all the following lines?

File: zookeeper-server/src/test/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProviderTest.java [hunk 6]
```diff
@@ -266,6 +266,20 @@ public void testConnectionFiltering() {
     provider
         .handleAuthentication(new ServerAuthenticationProvider.ServerObjs(zks, closableMockServerCnxn), new byte[0]);
     Assert.assertTrue(closableMockServerCnxn.isClosed());
+    System.clearProperty(X509AuthenticationConfig.DEDICATED_DOMAIN);
+
+    // Non-authorized user (connection filtering not enforced)
+    closableMockServerCnxn = new ClosableMockServerCnxn();
+    closableMockServerCnxn.clientChain = new X509Certificate[]{domainXCert};
+    System.setProperty(X509AuthenticationConfig.DEDICATED_DOMAIN, "DomainY");
+    System.setProperty(X509AuthenticationConfig.ENFORCE_DEDICATED_DOMAIN, "false");
```
Comment (desaikomal) [inline zookeeper-server/src/test/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProviderTest.java:275]: do you want to add any test cases for  Non-authorized user with connection filtering enforced? May be you already have that test case. Sorry, if it is already there and i didn't see

**Review by desaikomal — COMMENTED** [review]
If you can answer the test case related, thing, that will help to approve the change.

File: zookeeper-client/zookeeper-client-c/src/zookeeper.c (hunk 1 above)
Comment (abhilash1in) [reply zookeeper-client/zookeeper-client-c/src/zookeeper.c:2595]: This commit/change is required to unblock CI builds. See https://github.com/apache/zookeeper/pull/1951 .

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java (hunk 3 above)
Comment (abhilash1in) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java:382]: I'm following the getter/setter pattern used for rest of the variables.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java (hunk 4 above)
Comment (abhilash1in) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java:384]: Good catch, but [Boolean.parseBoolean()](https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html#parseBoolean-java.lang.String-) is case insensitive (has an implicit ignoreCase).

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeerConfig.java (hunk 5 above)
Comment (abhilash1in) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeerConfig.java:380]: The description of this PR has more details on this.

File: zookeeper-server/src/test/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProviderTest.java (hunk 6 above)
Comment (abhilash1in) [reply zookeeper-server/src/test/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProviderTest.java:275]: It's there already. Please see line 259 above.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/admin/Commands.java (hunk 2 above)
Comment (abhilash1in) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/admin/Commands.java]: good catch. The variable was used in another place at one point during debugging, but is no longer used/needed. I reverted it to a single line.

**Review by desaikomal — APPROVED** [review]
thanks for working on this.

**Review by desaikomal — APPROVED** [review]
thanks for working on this.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeerConfig.java (hunk 5 above)
Comment (rahulrane50) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeerConfig.java:380]: If this is not correct then essentially we are doing lazy binding of these variables. Other way to fix this is fix these if blocks. I would suggest separate out this change into separate PR and test it independently and make sure "all" these fields are lazily bound in their getter methods.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java (hunk 3 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java:382]: If you want to follow rest of the variables pattern then we should lazy bind it in getter method and have only one setter method. I suggest we should do it so that we can keep it consistent and it will also keep setter method names consistent.

**Review by rahulrane50 — COMMENTED** [review]
Overall LGTM thanks for making this change.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java (hunk 3 above)
Comment (abhilash1in) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java:382]: `shouldEnforceDedicatedDomain()` is the getter method and is already following the convention you mentioned. Am I missing something?
